lib/EarlyFunctionAnalysis/DetectABI.cpp

```diff
@@ -488,6 +494,99 @@ void DetectABI::interproceduralPropagation() {
   }
 }
 
+void DetectABI::propagatePrototypes() {
+  for (model::Function &Function : Binary->Functions()) {
+    const auto &Entry = Function.Entry();
+
+    auto &Summary = Oracle.getLocalFunction(Entry);
+    auto &ABI = Summary.ABIResults;
+    auto &CFG = Summary.CFG;
+    auto &WrittenRegisters = Summary.WrittenRegisters;
+
+    const bool HasNoArguments = ABI.ArgumentsRegisters.size() == 0;
+    const bool HasNoReturns = ABI.FinalReturnValuesRegisters.size() == 0;
+    const bool IsSingleNode = CFG.size() == 1;
+
+    if (HasNoReturns and HasNoArguments and IsSingleNode) {
+      auto &Block = *CFG.begin();
+
+      if (Block.Successors().size() == 1) {
+        auto &Successor = *Block.Successors().begin();
+
+        // Select new prototype for wrapper function
+        if (auto Call = dyn_cast<efa::CallEdge>(Successor.get())) {
+          auto Prototype = getPrototype(*Binary, Entry, Block, *Call);
```

Comment:
Turns this "precondition checks" into early loop `continue`.

```c++
if (not FirstCondition)
  continue;
if (not SecondCondition)
  continue;
if (not ThirdCondition)
  continue;

Reply:
You mean [this](https://github.com/revng/revng/pull/299#discussion_r1217965540)?

Outlining and having `return`s seems like a good thing to do.

Having a lot of nested `if`s is not great, a lot of `continue`s are sad too, outlining and using `return` is the best solution.